### ISSUE TITLE
Unable to execute external binaries in bin using zip deploy and Github Actions

### ISSUE THREAD

**Issue by vsub21** [Azure/azure-functions-host#6051]
I am running my function on the **Linux Consumption** plan (in Python) and I have an external binary that I have added to my project within the bin folder (a static binary for **ffmpeg**). Upon attempting to execute this program, I get a permission denied error. Even after changing the permissions on this file with chmod 777, I am unable to run this file and receive a permission denied error in the logs upon running my function.

I am deploying using Github Actions with this as a template (https://github.com/Azure/actions-workflow-samples/blob/master/FunctionApp/linux-python-functionapp-on-azure.yml) -- I have attempted adding a new job to perform chmod 777 to change permissions for the file as such:

```
    - name: 'Make ffmpeg executable'
      shell: bash
      run: |
        echo "Making executable the following file: ${{ env.FFMPEG_PATH }}"
        ls -la ${{ env.FFMPEG_PATH }}
        chmod 777 ${{ env.FFMPEG_PATH }}
        ls -la ${{ env.FFMPEG_PATH }}
```
with the following output upon action execution:

```
Run echo "Making executable the following file: bin/ffmpeg-git-20200504-amd64-static/ffmpeg"
Making executable the following file: bin/ffmpeg-git-20200504-amd64-static/ffmpeg
-rw-r--r-- 1 runner docker 74890952 May 17 16:18 bin/ffmpeg-git-20200504-amd64-static/ffmpeg
-rwxrwxrwx 1 runner docker 74890952 May 17 16:18 bin/ffmpeg-git-20200504-amd64-static/ffmpeg
```
I have even tried performing os.popen("ls -la").read() and os.popen("chmod 777 {FFMPEG_PATH}").read() within my python script to change the permissions of my file as such:

```
    # Give ffmpeg 777 perms

    chmod_before_string = "ls -la {}".format(str(sakuraitweetbot.FFMPEG_PATH))
    chmod_before_res = os.popen(chmod_before_string).read()

    logger.info('Executed: "{}"'.format(chmod_before_string))
    logger.info('Output: "{}"'.format(chmod_before_res))

    chmod_string = "chmod 777 {}".format(str(sakuraitweetbot.FFMPEG_PATH))
    chmod = os.popen(chmod_before_string).read()
    logger.info('Executed: "{}"'.format(chmod_string))

    chmod_after_string = "ls -la {}".format(str(sakuraitweetbot.FFMPEG_PATH))
    chmod_after_res = os.popen(chmod_after_string).read()

    logger.info('Executed: "{}"'.format(chmod_after_string))
    logger.info('Output: "{}"'.format(chmod_after_res))
```

The logs for this are as such:

```
2020-05-18T13:00:00.076676Z	Executed: "ls -la /home/site/wwwroot/<FUNCTION_NAME>/../bin/ffmpeg-git-20200504-amd64-static/ffmpeg"
2020-05-18T13:00:00.0811441Z	"Output: ""-rw-r--r-- 1 root root 74890952 May 17 16:18 /home/site/wwwroot/<FUNCTION_NAME>/../bin/ffmpeg-git-20200504-amd64-static/ffmpeg
"""
2020-05-18T13:00:00.0854194Z	Executed: "chmod 777 /home/site/wwwroot/<FUNCTION_NAME>/../bin/ffmpeg-git-20200504-amd64-static/ffmpeg"
2020-05-18T13:00:00.100643Z	Executed: "ls -la /home/site/wwwroot/<FUNCTION_NAME>/../bin/ffmpeg-git-20200504-amd64-static/ffmpeg"
2020-05-18T13:00:00.11Z	"Output: ""-rw-r--r-- 1 root root 74890952 May 17 16:18 /home/site/wwwroot/<FUNCTION_NAME>/../bin/ffmpeg-git-20200504-amd64-static/ffmpeg
"""
```

Even after performing chmod using the os module, the execute permission is not assigned.

Is this because of the mounting to a read-only filesystem, and is there a workaround to fix this?

**Comment by vsub21** [comment]
@brettsam Gentle bump on this -- have two minutes to take a look at this issue? 😄

**Comment by balag0** [comment]
just to verify if this is due to the read only fs, can you try copying the file to the temp folder and run it from there?
https://docs.microsoft.com/en-us/azure/azure-functions/functions-reference-python#temporary-files

**Comment by vsub21** [comment]
@balag0 Just tried it this morning. I modified the my Python code to copy ```ffmpeg``` to ```/tmp``` as such:

```
    # Copy ffmpeg to tmp and give 777 perms
    TMP_FFMPEG_PATH = "/tmp/ffmpeg"

    copy_file_string = "cp -v {} {}".format(str(sakuraitweetbot.FFMPEG_PATH), TMP_FFMPEG_PATH)
    copy_file_res = os.popen(copy_file_string).read()

    logger.info('Executed: "{}"'.format(copy_file_string))
    logger.info('Output: "{}"'.format(copy_file_res))

    chmod_before_string = "ls -la {}".format(TMP_FFMPEG_PATH)
    chmod_before_res = os.popen(chmod_before_string).read()

    logger.info('Executed: "{}"'.format(chmod_before_string))
    logger.info('Output: "{}"'.format(chmod_before_res))

    chmod_string = "chmod 777 {}".format(TMP_FFMPEG_PATH)
    chmod = os.popen(chmod_before_string).read()
    logger.info('Executed: "{}"'.format(chmod_string))

    chmod_after_string = "ls -la {}".format(TMP_FFMPEG_PATH)
    chmod_after_res = os.popen(chmod_after_string).read()

    logger.info('Executed: "{}"'.format(chmod_after_string))
    logger.info('Output: "{}"'.format(chmod_after_res))
```
I got the following in my logs:

```
2020-05-27T16:51:49.9282343Z | Output:   "'/home/site/wwwroot/<FUNCTION_NAME>/../bin/ffmpeg-git-20200504-amd64-static/ffmpeg'   -> '/tmp/ffmpeg'     "
-- | --
2020-05-27T16:51:49.9410573Z | Executed: "ls -la /tmp/ffmpeg"
2020-05-27T16:51:49.9478927Z | Output: "-rw-r--r-- 1 app   app 74890952 May 27 16:51 /tmp/ffmpeg     "
2020-05-27T16:51:49.9547295Z | Executed: "chmod 777 /tmp/ffmpeg"
2020-05-27T16:51:49.958526Z | Executed: "ls -la /tmp/ffmpeg"
2020-05-27T16:51:49.9657683Z | Output: "-rw-r--r-- 1 app   app 74890952 May 27 16:51 /tmp/ffmpeg     "
```
It does not seem to have changed the permissions as expected. [Here](https://github.com/vsub21/sakurai-tweetbot/blob/master/sakuraitweetbot-function/__init__.py) is a link to the ```__init__.py``` and the project. Any ideas?

Also -- from what I understand, the contents of ```/tmp``` do not persist across each run (function is supposed to be stateless). Having to copy this file to ```/tmp``` may not be the best solution -- are there are ways to accomplish this with the Linux Consumption plan? 

I am also using the zip deploy method since the integration with Github Actions makes it easy to auto-deploy whenever I ```git push```. Are there alternative deploy methods I should perhaps look at?

**Comment by balag0** [comment]
You are right, files in temp do not persist across instances. This was just an experiment to see if that helps with modifying file permissions.

I found this earlier issue where executable permissions were not being retained in some scenarios.
https://github.com/Azure/azure-functions-core-tools/issues/1612

Based on this it looks like if you use "--build-native-deps" flag when publishing to the function app using functions core tools it should work. Can you please give it a try?

https://docs.microsoft.com/en-us/azure/azure-functions/functions-run-local?tabs=windows%2Ccsharp%2Cbash#publish

**Comment by brettsam** [comment]
@vsub21 -- were you ever able to give @balag0's suggestion a try?

**Comment by ghost** [comment]
This issue has been automatically marked as stale because it has been marked as requiring author feedback but has not had any activity for **4 days**. It will be closed if no further activity occurs **within 3 days of this comment**.

**Comment by vsub21** [comment]
> @vsub21 -- were you ever able to give @balag0's suggestion a try?

Hi @brettsam, I found a workaround that allowed me to accomplish my goal without the dependency on ```ffmpeg``` itself. I did not yet test this particular scenario out using the Azure CLI, so the read/write/executable permissions may still be an issue with Github Actions deployment to Azure Functions. If I get a chance to test it out in the future, I'll update here.
